common/src/ess.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+use std::{sync::Arc, time::SystemTime};
+
+use crate::proto::{
+    common::Value as ValueMessage,
+    common::{value::Value as ValueEnum, SubscribeFulfillment, SubscribeIntent},
+    streaming::{channel_service_server::ChannelService, Event, OpenRequest},
+};
+use async_trait::async_trait;
+use ess::EventSubSystem;
+use tokio::spawn;
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::{Response, Status};
+use uuid::Uuid;
+
+type InnerEss<T> = EventSubSystem<Box<str>, Box<str>, T, Result<Event, Status>>;
+
+/// [`SharedEss`](SharedEss) integrates the reusable
+/// [`EventSubSystem`](EventSubSystem) component with the Chariott gRPC
+/// streaming contract. Cloning [`SharedEss`](SharedEss) is cheap, it will not
+/// create a new instance but refer to the same underlying instance instead.
+#[derive(Clone)]
+pub struct SharedEss<T>(Arc<InnerEss<T>>);
+
+impl<T: Clone> SharedEss<T> {
+    pub fn new() -> Self {
+        Self(Arc::new(EventSubSystem::new()))
+    }
+}
+
+impl<T: Clone> Default for SharedEss<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: Clone + Send + 'static> SharedEss<T> {
+    pub fn serve_subscriptions(
+        &self,
+        subscribe_intent: SubscribeIntent,
+        into_value: fn(T) -> ValueEnum,
+    ) -> Result<SubscribeFulfillment, Status> {
+        let subscriptions = self
+            .0
+            .register_subscriptions(
+                subscribe_intent.channel_id.into(),
+                subscribe_intent.sources.into_iter().map(|s| s.into()),
+            )
+            .map_err(|_| Status::failed_precondition("The specified client does not exist."))?;
+
+        for subscription in subscriptions {
+            let source = subscription.event_id().to_string();
+
+            spawn(subscription.serve(move |data, seq| {
+                Ok(Event {
+                    source: source.clone(),
+                    value: Some(ValueMessage { value: Some(into_value(data)) }),
+                    seq,
+                    timestamp: Some(SystemTime::now().into()),
+                })
+            }));
+        }
+
+        Ok(SubscribeFulfillment {})
+    }
+}
+
+#[async_trait]
+impl<T> ChannelService for SharedEss<T>
+where
+    T: Clone + Send + Sync + 'static,
+{
+    type OpenStream = ReceiverStream<Result<Event, Status>>;
+
+    async fn open(
+        &self,
+        _: tonic::Request<OpenRequest>,
+    ) -> Result<Response<Self::OpenStream>, Status> {
+        const METADATA_KEY: &str = "x-chariott-channel-id";
+
+        let id = Uuid::new_v4().to_string();
+        let (_, receiver_stream) = self.0.read_events(id.clone().into());
+        let mut response = Response::new(receiver_stream);
+        response.metadata_mut().insert(METADATA_KEY, id.try_into().unwrap());
+        Ok(response)
+    }
+}
+
+impl<T> AsRef<InnerEss<T>> for SharedEss<T> {
```

Comment:
I think what you want to implement here is `Deref` instead. Then you don't have to see all those `.as_ref()` calls on the call sites (in other words, from the caller's perspective, the wrapper appears to be like its inner type):

```diff
diff --git a/common/src/ess.rs b/common/src/ess.rs
index 662aa47..0c470a7 100644
--- a/common/src/ess.rs
+++ b/common/src/ess.rs
@@ -1,7 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT license.
 
-use std::{sync::Arc, time::SystemTime};
+use std::{ops::Deref, sync::Arc, time::SystemTime};
 
 use crate::proto::{
     common::Value as ValueMessage,
@@ -88,8 +88,10 @@ where
     }
 }
 
-impl<T> AsRef<InnerEss<T>> for SharedEss<T> {
-    fn as_ref(&self) -> &InnerEss<T> {
+impl<T> Deref for SharedEss<T> {
+    type Target = InnerEss<T>;
+
+    fn deref(&self) -> &Self::Target {
         self.0.as_ref()
     }
 }
diff --git a/examples/common/src/chariott/streaming.rs b/examples/common/src/chariott/streaming.rs
index 99ae11a..79e4250 100644
--- a/examples/common/src/chariott/streaming.rs
+++ b/examples/common/src/chariott/streaming.rs
@@ -18,7 +18,7 @@ struct Ess<T>(InnerEss<(EventId, T)>);
 
 impl<T: Clone + Send + 'static> Observer<EventId, T> for Ess<T> {
     fn on_set(&mut self, key: &EventId, value: &T) {
-        self.0.as_ref().publish(key, (key.clone(), value.clone()));
+        self.0.publish(key, (key.clone(), value.clone()));
     }
 }
 
diff --git a/src/registry.rs b/src/registry.rs
index 54b75bf..9917f4f 100644
--- a/src/registry.rs
+++ b/src/registry.rs
@@ -36,7 +36,7 @@ impl Observer for Ess {
             })
             .collect::<HashSet<_>>()
         {
-            self.as_ref().publish(format!("namespaces[{}]", namespace).as_str(), ());
+            self.publish(format!("namespaces[{}]", namespace).as_str(), ());
         }
     }
 }
```

Reply:
Reconsidering this: that the `SharedEss` _is_ not a smart pointer but _contains_ a smart pointer is a private implementation detail. Seen from the outside it still behaves as if it were a smart pointer, hence I applied your suggestion with 11c04e890f14980efa2adb5137c2b364624ff681.